### PR TITLE
[#466] #IMPLEMENT 'assemblyName: DotNet.Testcontainers; function: Ima…

### PR DESCRIPTION
**Pull request by bohlenc**
…ge Name Substitution'

### CODE REVIEW

File: tests/DotNet.Testcontainers.Tests/Unit/Builders/TestcontainerBuilderTest.cs [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  [InlineData("my.proxy.com:443", "bar:1.0.0", "my.proxy.com:443/bar:1.0.0")]
+  [InlineData("my.proxy.com", "foo/bar:1.0.0", "my.proxy.com/foo/bar:1.0.0")]
+  [InlineData("my.proxy.com/my-path", "foo/bar:1.0.0", "my.proxy.com/my-path/foo/bar:1.0.0")]
+  [InlineData("my.proxy.com:443", "foo/bar:1.0.0", "my.proxy.com:443/foo/bar:1.0.0")]
+  [InlineData("my.proxy.com:443/my-path", "foo/bar:1.0.0", "my.proxy.com:443/my-path/foo/bar:1.0.0")]
+  [InlineData("my.proxy.com", "myregistry.azurecr.io/foo/bar:1.0.0", "myregistry.azurecr.io/foo/bar:1.0.0")]
+  [InlineData("my.proxy.com", "myregistry.azurecr.io:443/foo/bar:1.0.0", "myregistry.azurecr.io:443/foo/bar:1.0.0")]
+  public void ImageNameFromStringIsPrefixedWhenPrefixIsSet(string prefix, string originalImageName, string expectedName)
+  {
+    TestcontainersSettings.DockerHubImagePrefix = prefix;
```
Comment (HofmeisterAn) [inline tests/DotNet.Testcontainers.Tests/Unit/Builders/TestcontainerBuilderTest.cs]: We just need to make sure, that the tests won't run with other tests. Changing the static property might not be a good idea with parallel tests.

**Review by HofmeisterAn — COMMENTED** [review]
Thanks. I think thats much cleaner.

File: tests/DotNet.Testcontainers.Tests/Unit/Builders/TestcontainerBuilderTest.cs (hunk 1 above)
Comment (bohlenc) [reply tests/DotNet.Testcontainers.Tests/Unit/Builders/TestcontainerBuilderTest.cs]: I wasn't aware tests are executed in parallel.

Fixed.

Comment (HofmeisterAn) [reply tests/DotNet.Testcontainers.Tests/Unit/Builders/TestcontainerBuilderTest.cs]: That's not how it works :). You'll need a `CollectionDefinition` first and then set `DisableParallelization`. I can add it. See: https://github.com/xunit/xunit/issues/1999#issuecomment-522635397.

Comment (bohlenc) [reply tests/DotNet.Testcontainers.Tests/Unit/Builders/TestcontainerBuilderTest.cs]: Then maybe the xunit documentation needs fixing: https://xunit.net/docs/running-tests-in-parallel > "Custom Test Collections". Other tests relying on `TestcontainersSettings.DockerHubImagePrefix` then just would have to be in the same collection, no?

Or am I missing something there?

Comment (bohlenc) [reply tests/DotNet.Testcontainers.Tests/Unit/Builders/TestcontainerBuilderTest.cs]: > That's not how it works :). You'll need a `CollectionDefinition` first and then set `DisableParallelization`. I can add it. See: [xunit/xunit#1999 (comment)](https://github.com/xunit/xunit/issues/1999#issuecomment-522635397).

Yep, saw that issue. From what I gather from the docs though that is not required. Maybe I'm wrong.

Comment (HofmeisterAn) [reply tests/DotNet.Testcontainers.Tests/Unit/Builders/TestcontainerBuilderTest.cs]: Each test class is a unique test collection. By default, test collections run parallel. The unit tests in this library are in different test collections. Tests that access the static resource (almost all of them) can't run parallel.

> Other tests relying on `TestcontainersSettings.DockerHubImagePrefix` then just would have to be in the same collection, no?
> 
> Or am I missing something there?

Yes, exactly. Almost all tests rely on that static property. But putting all of them in the same collection is way too complicated. We can just take advantage of the `[CollectionDefinition(DisableParallelization = true)]` behavior:

> Parallel-capable test collections will be run first (in parallel), followed by parallel-disabled test collections (run sequentially).

Comment (bohlenc) [reply tests/DotNet.Testcontainers.Tests/Unit/Builders/TestcontainerBuilderTest.cs]: Ah, got it. Thanks!

Comment (bohlenc) [reply tests/DotNet.Testcontainers.Tests/Unit/Builders/TestcontainerBuilderTest.cs]: I added a simple `ParallelizationDisabled` collection.

**Review by HofmeisterAn — APPROVED** [review]
I'll merge it later that day. I'll add a minor change that changes the `IDockerContainer.ImageName` return type to `IDockerImage` and reduces the amount of `InlineData`, we can replace it with `MemberData` (incl. the code smell).

**Comment by HofmeisterAn** [comment]
Published in `1.6.0-beta.2242`, closes #466.

**Pull request closed without merging**